### PR TITLE
 feat(connector): [Braintree] implement dispute webhook 

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

1.Dispute Webhook Support for Braintree:
2.Additional Field for MerchantConnectorWebhookDetails:
Introduced a new field named additional_secret within the MerchantConnectorWebhookDetails struct to accommodate additional data.
3.Enhanced Source Verification Signature Retrieval:
Modified the get_webhook_source_verification_signature() function to include an additional parameter named _secret for improved source verification.

## How did you test it?

Screenshot 2023-08-28 at 12 55 32 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/types/api/webhooks.rs [hunk 1]
```diff
@@ -106,43 +106,50 @@ pub trait IncomingWebhook: ConnectorCommon + Sync {
             )
             .await;
 
-        let merchant_secret = match merchant_connector_account_result {
+        let (merchant_secret, additional_secret) = match merchant_connector_account_result {
```
Comment (ArjunKarthik) [inline crates/router/src/types/api/webhooks.rs:109]: Instead of tuple use type which will have 2 fields

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 2]
```diff
@@ -982,3 +983,79 @@ impl<F, T>
...
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct TestResponse {
+    pub timestamp: Option<String>,
+}
+#[derive(Debug, Deserialize, Serialize)]
+pub struct BraintreeDisputeData {
+    pub amount_disputed: i64,
```
[the remark this answers is not in the document]
Comment (AkshayaFoiger) [reply crates/router/src/connector/braintree/braintree_graphql_transformers.rs:1022]: They send amount in dollars. Modified the response to convert dollars into cents.

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 3]
```diff
@@ -982,3 +983,79 @@ impl<F, T>
...
+            "dispute_auto_accepted" => Self::DisputeAccepted,
+            "dispute_expired" => Self::DisputeExpired,
+            "dispute_disputed" => Self::DisputeChallenged,
+            _ => Self::EventNotSupported,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct TestResponse {
```
[the remark this answers is not in the document]
Comment (AkshayaFoiger) [reply crates/router/src/connector/braintree/braintree_graphql_transformers.rs]: Removed TestResponse struct

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by prasunna09 — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by prasunna09 — DISMISSED** [review]

File: crates/router/src/types/api/webhooks.rs [hunk 4]
```diff
@@ -126,6 +126,7 @@ pub trait IncomingWebhook: ConnectorCommon + Sync {
     fn get_webhook_source_verification_signature(
         &self,
         _request: &IncomingWebhookRequestDetails<'_>,
+        _connector_webhook_secrets: &api_models::webhooks::ConnectorWebhookSecrets,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/types/api/webhooks.rs:129]: why is this change required? you can always access `get_webhook_source_verification_merchant_secret` in `get_webhook_source_verification_signature` using self. Currently get_webhook_source_verification_merchant_secret is async, you can change it to sync fn to accomodate this change!

File: crates/router/src/types/api/webhooks.rs [hunk 5]
```diff
@@ -134,7 +135,7 @@ pub trait IncomingWebhook: ConnectorCommon + Sync {
         &self,
         _request: &IncomingWebhookRequestDetails<'_>,
         _merchant_id: &str,
-        _secret: &[u8],
+        _connector_webhook_secrets: &api_models::webhooks::ConnectorWebhookSecrets,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/types/api/webhooks.rs:138]: you can always access `get_webhook_source_verification_merchant_secret` in `get_webhook_source_verification_message`

**Review by sai-harsha-vardhan — CHANGES_REQUESTED** [review]

File: crates/router/src/types/api/webhooks.rs (hunk 4 above)
Comment (AkshayaFoiger) [reply crates/router/src/types/api/webhooks.rs:129]: `MerchantAccount` and `MerchantConnectorAccount` are necessary parameters for invoking `get_webhook_source_verification_merchant_secret` but the scope of these parameters are not extended to `get_webhook_source_verification_signature`

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 6]
```diff
@@ -1411,3 +1412,75 @@ fn get_braintree_redirect_form(
...
+    pub dispute: Option<BraintreeDisputeData>,
+}
+impl types::transformers::ForeignFrom<&str> for api_models::webhooks::IncomingWebhookEvent {
+    fn foreign_from(status: &str) -> Self {
+        match status {
+            "dispute_opened" => Self::DisputeOpened,
+            "dispute_lost" => Self::DisputeLost,
+            "dispute_won" => Self::DisputeWon,
+            "dispute_accepted" => Self::DisputeAccepted,
+            "dispute_auto_accepted" => Self::DisputeAccepted,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/braintree/braintree_graphql_transformers.rs:1436]:
```suggestion
"dispute_accepted" |  "dispute_auto_accepted" => Self::DisputeAccepted,
```

File: crates/router/src/connector/braintree/braintree_graphql_transformers.rs [hunk 7]
```diff
@@ -1411,3 +1412,75 @@ fn get_braintree_redirect_form(
...
+    pub bt_signature: String,
+    pub bt_payload: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct Notification {
+    pub kind: String, // xml parse only string to fields
+    pub timestamp: String,
+    pub dispute: Option<BraintreeDisputeData>,
```
[the remark this answers is not in the document]
Comment (AkshayaFoiger) [reply crates/router/src/connector/braintree/braintree_graphql_transformers.rs:1427]: There are payment webhooks, for some payment methods like ACH, SEPA and the response is given in XML format. That doesn't support enum de-serialization. So object Notification will be common for all webhooks, therefore we have made dispute an optional field.

**Review by ArjunKarthik — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]
